### PR TITLE
Internal: Allow hosting to replace promotions - support kit library [ED-13801]

### PR DESCRIPTION
**Pull request by elchugreeva**

### CODE REVIEW

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: app/modules/kit-library/module.php [hunk 1]
```diff
@@ -75,6 +78,20 @@ private function set_kit_library_settings() {
 		] );
 	}
 
+	private function apply_filter_subscription_plans( array $subscription_plans ): array {
+		$plans = array_keys( $subscription_plans );
+
+		foreach ( $plans as $plan ) {
+			$subscription_plans[ $plan ] = Filtered_Promotions_Manager::get_filtered_promotion_data(
```
Comment (TzviRabinovitch) [inline app/modules/kit-library/module.php:85]: skip if 'promotion_url' == null

File: app/modules/kit-library/module.php [hunk 2]
```diff
@@ -75,6 +78,21 @@ private function set_kit_library_settings() {
 		] );
 	}
 
+	private function apply_filter_subscription_plans( array $subscription_plans ): array {
+		foreach ( $subscription_plans as $key => $plan ) {
+			if ( $plan[ 'promotion_url' ] == null) {
+				continue;
+			}
+			$subscription_plans[ $key ] = Filtered_Promotions_Manager::get_filtered_promotion_data(
```
Comment (louiswol94) [inline app/modules/kit-library/module.php:86]:
```suggestion
			if ( null === $plan[ 'promotion_url' ] ) {
				continue;
			}

			$subscription_plans[ $key ] = Filtered_Promotions_Manager::get_filtered_promotion_data(
```
Or wrap everything in `if( isset( $plan[ 'promotion_url' ] ) ) { ... }`

File: app/modules/kit-library/module.php [hunk 3]
```diff
@@ -47,6 +48,7 @@ private function register_admin_menu_legacy( Admin_Menu_Manager $admin_menu ) {
 		);
 	}
 
+
```
Comment (louiswol94) [inline app/modules/kit-library/module.php:51]:
```suggestion
```

File: app/modules/kit-library/module.php [hunk 4]
```diff
@@ -56,11 +58,12 @@ private function set_kit_library_settings() {
 		$connect = Plugin::$instance->common->get_component( 'connect' );
 
 		/** @var Kit_Library $kit_library */
+
```
Comment (louiswol94) [inline app/modules/kit-library/module.php:61]:
```suggestion
```

File: app/modules/kit-library/module.php [hunk 5]
```diff
@@ -75,6 +78,20 @@ private function set_kit_library_settings() {
 		] );
 	}
 
+	private function apply_filter_subscription_plans( array $subscription_plans ): array {
+		$plans = array_keys( $subscription_plans );
+
+		foreach ( $plans as $plan ) {
```
[the remark this answers is not in the document]
Comment (davseve) [reply app/modules/kit-library/module.php]: `$plans` is a flat array.
`[ "free", "pro", "essential", ....]`

File: app/modules/kit-library/module.php [hunk 6]
```diff
@@ -75,6 +78,21 @@ private function set_kit_library_settings() {
 		] );
 	}
 
+	private function apply_filter_subscription_plans( array $subscription_plans ): array {
+		foreach ( $subscription_plans as $key => $plan ) {
+			if ($plan['promotion_url'] == null) {
```
Comment (davseve) [inline app/modules/kit-library/module.php:83]:
```suggestion
			if ( null === $plan['promotion_url'] ) {
```

File: app/modules/kit-library/module.php [hunk 7]
```diff
@@ -46,7 +47,7 @@ private function register_admin_menu_legacy( Admin_Menu_Manager $admin_menu ) {
 			new Kit_Library_Menu_Item()
 		);
 	}
-
+
```
Comment (louiswol94) [inline app/modules/kit-library/module.php:50]:
```suggestion

```

File: app/modules/kit-library/module.php [hunk 8]
```diff
@@ -46,7 +47,6 @@ private function register_admin_menu_legacy( Admin_Menu_Manager $admin_menu ) {
 			new Kit_Library_Menu_Item()
 		);
 	}
-
 	private function set_kit_library_settings() {
```
Comment (louiswol94) [inline app/modules/kit-library/module.php:50]:
```suggestion

	private function set_kit_library_settings() {
```

File: app/modules/kit-library/module.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (louiswol94) [reply app/modules/kit-library/module.php]: @davseve Are you sure?

`array(6) { ["free"]=> array(3) { ["label"]=> NULL ["promotion_url"]=> NULL ["color"]=> NULL } ["essential"]=> array(3) { ["label"]=> string(3) "Pro" ["promotion_url"]=> string(94) "https://elementor.com/pro/?utm_source=kit-library&utm_medium=wp-dash&utm_campaign=gopro&louis1" ["color"]=> string(7) "#92003B" } ["essential-oct2023"]=> array(3) { ["label"]=> string(8) "Advanced" ["promotion_url"]=> string(94) "https://elementor.com/pro/?utm_source=kit-library&utm_medium=wp-dash&utm_campaign=gopro&louis2" ["color"]=> string(7) "#92003B" } ["advanced"]=> array(3) { ["label"]=> string(8) "Advanced" ["promotion_url"]=> string(94) "https://elementor.com/pro/?utm_source=kit-library&utm_medium=wp-dash&utm_campaign=gopro&louis3" ["color"]=> string(7) "#92003B" } ["expert"]=> array(3) { ["label"]=> string(6) "Expert" ["promotion_url"]=> string(94) "https://elementor.com/pro/?utm_source=kit-library&utm_medium=wp-dash&utm_campaign=gopro&louis4" ["color"]=> string(7) "#92003B" } ["agency"]=> array(3) { ["label"]=> string(6) "Agency" ["promotion_url"]=> string(94) "https://elementor.com/pro/?utm_source=kit-library&utm_medium=wp-dash&utm_campaign=gopro&louis5" ["color"]=> string(7) "#92003B" } }`

[the remark this answers is not in the document]
Comment (davseve) [reply app/modules/kit-library/module.php]: @louiswol94 You have pasted the `$subscription_plans`, 
The `$plans` it the outcome of `array_keys( $subscription_plans );`

**Review by louiswol94 — DISMISSED** [review]

File: app/modules/kit-library/module.php [hunk 9]
```diff
@@ -75,6 +76,22 @@ private function set_kit_library_settings() {
...
 
+	private function apply_filter_subscription_plans( array $subscription_plans ): array {
+		foreach ( $subscription_plans as $key => $plan ) {
+			if ( null === $plan['promotion_url'] ) {
+				continue;
+			}
+
+			$subscription_plans[ $key ] = Filtered_Promotions_Manager::get_filtered_promotion_data(
+				$plan,
+				'elementor/kit_library/'. $key . '/promotion',
```
Comment (louiswol94) [inline app/modules/kit-library/module.php:87]:
```suggestion
				'elementor/kit_library/' . $key . '/promotion',
```

**Review by louiswol94 — APPROVED** [review]

**Pull request merged**